akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -432,17 +435,21 @@ import scala.util.control.NonFatal
     eventsRemaining
   }
 
-  def runAsyncInput(logic: GraphStageLogic, evt: Any, handler: (Any) ⇒ Unit): Unit =
+  def runAsyncInput(logic: GraphStageLogic, evt: Any, promise: OptionVal[Promise[Done]], handler: (Any) ⇒ Unit): Unit =
     if (!isStageCompleted(logic)) {
       if (GraphInterpreter.Debug) println(s"$Name ASYNC $evt ($handler) [$logic]")
       val currentInterpreterHolder = _currentInterpreter.get()
       val previousInterpreter = currentInterpreterHolder(0)
       currentInterpreterHolder(0) = this
       try {
         activeStage = logic
-        try handler(evt)
-        catch {
-          case NonFatal(ex) ⇒ logic.failStage(ex)
+        try {
+          handler(evt)
+          if (promise.isDefined) promise.get.success(Done)
+        } catch {
+          case NonFatal(ex) ⇒
+            if (promise.isDefined) promise.get.failure(ex)
```

Comment:
and `failStage` will do `tryFailure` (StreamDetached)?

Reply:
`failStage` does nothing with the promise, but later `ConcurrentAsyncCallback.afterPostStop` is called, and that will `tryFailure` any outstanding uncompleted feedback-promises. So this is a fail-fast to keep the exception when it was easy to capture it. 

It could perhaps be nicer to some how add a trigger from inside `failStage` instead and always keep the exception. WDYT?